ports/stm32f4/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,426 @@
+    if (clk_div != 0) {
+        // APB prescaler for this timer is > 1
+        source *= 2;
+    }
+    return source;
+}
+
+void pwmout_reset(void) {
+    uint16_t never_reset_mask = 0x00;
+    for(int i=0;i<TIM_BANK_ARRAY_LEN;i++) {
+        if (!never_reset_tim[i]) {
+            reserved_tim[i] = 0x00;
+            tim_frequencies[i] = 0x00;
+        } else {
+            never_reset_mask |= 1<<i;
+        }
+    }
+    tim_clock_disable(ALL_CLOCKS & ~(never_reset_mask));
+}
+
+void common_hal_pulseio_pwmout_never_reset(pulseio_pwmout_obj_t *self) {
+    for(size_t i = 0 ; i < TIM_BANK_ARRAY_LEN; i++) {
+        if (mcu_tim_banks[i] == self->handle.Instance) {
+            never_reset_tim[i] = true;
+            never_reset_pin_number(self->tim->pin->port, self->tim->pin->number);
+            break;
+        }
+    }
+}
+
+void common_hal_pulseio_pwmout_reset_ok(pulseio_pwmout_obj_t *self) {
+    for(size_t i = 0 ; i < TIM_BANK_ARRAY_LEN; i++) {
+        if (mcu_tim_banks[i] == self->handle.Instance) {
+            never_reset_tim[i] = false;
+            break;
+        }
+    }
+}
+
+pwmout_result_t common_hal_pulseio_pwmout_construct(pulseio_pwmout_obj_t* self,
+                                                    const mcu_pin_obj_t* pin,
+                                                    uint16_t duty,
+                                                    uint32_t frequency,
+                                                    bool variable_frequency) {
+    TIM_TypeDef * TIMx;
+    uint8_t tim_num = sizeof(mcu_tim_pin_list)/sizeof(*mcu_tim_pin_list);
+    bool tim_chan_taken = false;
+    bool tim_taken_f_mismatch = false;
+    bool var_freq_mismatch = false;
+    bool first_time_setup = true;
+
+    for(uint i = 0; i < tim_num; i++) {
+        //if pin is same
+        if (mcu_tim_pin_list[i].pin == pin) {
+            //check if the timer has a channel active
+            if (reserved_tim[mcu_tim_pin_list[i].tim_index-1] != 0) {
+                //is it the same channel? (or all channels reserved by a var-freq)
+                if (reserved_tim[mcu_tim_pin_list[i].tim_index-1] & 1<<(mcu_tim_pin_list[i].channel_index-1)) {
+                    tim_chan_taken = true;
+                    continue; //keep looking, might be another viable option
+                }
+                //If the frequencies are the same it's ok
+                if (tim_frequencies[mcu_tim_pin_list[i].tim_index-1] != frequency) {
+                    tim_taken_f_mismatch = true;
+                    continue; //keep looking
+                }
+                //you can't put a variable frequency on a partially reserved timer
+                if (variable_frequency) {
+                    var_freq_mismatch = true;
+                    continue; //keep looking
+                }
+                first_time_setup = false; //skip setting up the timer
+            }
+            //No problems taken, so set it up
+            self->tim = &mcu_tim_pin_list[i];
```

Comment:
Please add intermediate variables for `mcu_tim_pin_list[i]`, `mcu_tim_pin_list[i].tim_index-1` and `reserved_tim[mcu_tim_pin_list[i].tim_index-1]`. These are used in multiple spots here and naming them with variables will make it clearer.

Reply:
Let me know how you feel about my reductions. I felt it was a little clearer with reserved_tim still in there but with better indexes

Reply:
let it be known that here lies the cursed change.

Reply:
Please elaborate. This comment has no context.

Reply:
Sorry, this wasn't a well thought out comment. I wanted to make a note for myself (in my comment history) of where this happened in the review process - the introduction of the intermediate variable for `mcu_tim_pin_list[i]` here caused a dangling pointer that had a cascade of bad effects, and I'd definitely like to make sure I don't make such a glaring error in the future, so I wanted to put in a reminder I could search for. Not professional wording though, I apologize.

Reply:
I'm not worried about the wording. It would be great to write up a full explanation of what the bug was and how this caused it. That way other folks can follow along and learn too.

Reply:
Sure thing. A dangling pointer was created by attempting to use intermediate variables to increase readability in this section, by creating a copy of the static constant list, `mcu_tim_pin_list[i]`, in an intermediate variable called l_tim. The `self->tim` member was then based off the l_tim copy, which then dropped out of scope and caused errors later in the program when essential pin information was overwritten without notice. I should have noticed the repercussions this would have before implementing the suggestion, given that I put together the peripheral tables in this way - next time I'll need to be more careful about my pointer management in construct functions.

Reply:
Thanks! I look forward to a PR fixing it.